tools/DeviceExplorer/DeviceExplorer/MainForm.cs

```diff
@@ -1102,5 +1103,71 @@ private void devicesGridViewContextMenu_Opened(object sender, EventArgs e)
                 showDevicePropertiesToolStripMenuItem.Enabled = false;
             }
         }
+
+        #region Device Filtering
+        private void filterDevicesTextBox_TextChanged(object sender, EventArgs e)
+        {
+            filterDevices();
+        }
+
+        private void filterDevices() { filterDevices(filterDevicesTextBox.Text); }
+        private void filterDevices(string filterText)
+        {
+            Debug.WriteLine($"Filtering on {filterText}");
+            if (!IsValidRegex(filterText))
+            {
+                Debug.WriteLine($"'\t{filterText}' is not a valid regex");
+                devicesGridView.DataSource = allDevices;
+                return;
+            }
+
+            IEnumerable<DeviceEntity> filteredDevices = 
+                from device in allDevices
+                where DeviceMatchesFilterText(device, filterText)
+                select device;
+
+            IEnumerable<DeviceEntity> deviceEntities = filteredDevices as DeviceEntity[] ?? filteredDevices.ToArray();
+            Debug.WriteLine($"Filtered down to {deviceEntities.Count()} devices out of {allDevices.Count}");
+
+            devicesGridView.DataSource = new SortableBindingList<DeviceEntity>(deviceEntities);
+        }
+
+        private static bool DeviceMatchesFilterText(DeviceEntity device, string filterText)
+        {
+            Debug.WriteLine($"\tChecking if device {device.Id} matches filter {filterText}");
+            return new[]
+            {
+                device.Id,
+                device.PrimaryKey,
+                device.SecondaryKey,
+                device.ConnectionState,
+                device.State,
+
+            }.Any(s => StringMatchesFilterText(s, filterText));
+        }
+
+        private static bool StringMatchesFilterText(string s, string filterText)
+        {
+            Debug.WriteLine($"\t\tChecking if string {s} matches filter {filterText}");
```

Comment:
Some of the properties can and will be `null`. E.g. Keys can be null if the device was created by the new Azure IoT Provisioning Service.

Adding the following fixes the issue:
`if (string.IsNullOrWhiteSpace(s)) return false;`

Reply:
Thank you for finding this!  This has been implemented!